server/src/watchdog.js

```diff
@@ -107,14 +107,53 @@ exports.submit = function(shot) {
   const imageId = imageUrl.split("/").pop();
   const form = new FormData();
 
-  Shot.getRawBytesForClip(
-    imageId
-  ).then((obj) => {
-    if (obj === null) {
-      const err = new Error("Unable to fetch image for Watchdog submission.");
-      err.imageId = String(imageId);
-      throw err;
-    }
+  // When this is called the shot has been successfully saved to the database,
+  // but the image data is not necessarily on S3 as the save to S3 is not
+  // awaited. Awaiting the S3 upload will affect the responsiveness of saving a
+  // shot with the addon.
+  //
+  // We'll give the S3 upload a head start and try a few times to fetch the
+  // image data. This is called for every shot saved, so let's keep it simple
+  // and minimal.
+
+  const getImageRawBytes = function() {
+    const MAX_ATTEMPTS = 3;
+    const DELAY = 1000; // milliseconds
+    let attemptCount = 1;
+
+    const delay = function(waitInMs) {
+      return new Promise((resolve) => {
+        setTimeout(() => resolve(), waitInMs);
+      });
+    };
+    const getRawBytesPromise = function(timeToWait) {
+      return delay(timeToWait).then(() => Shot.getRawBytesForClip(imageId));
+    };
+
+    return new Promise((resolve, reject) => {
+      const tryGetBytes = function() {
+        getRawBytesPromise(DELAY).then(obj => {
+          if (obj === null) {
```

Comment:
The beauty of Promise code is that it's not necessary to manually maintain a counter.

What if `getRawBytesPromise` just did something like:

```
// Three attempts with 1000 msec delay
getRawBytesPromise(1000)
  .then(bytes => { return bytes ? bytes : getRawBytesPromise(1000) })
  .then(bytes => { return bytes ? bytes : getRawBytesPromise(1000) })
  .then(bytes => { bytes ? resolve(bytes) : reject(new Error(...)) }
```

I think this approach would be simpler and more idiomatic.

Reply:
Er, sorry, I meant "what if `getImageRawBytes` did something like..."

Reply:
I'd argue in this case the counter is not necessary because we call the function three times.  Promises don't have anything to do with it.  😃  It is easier to read though.

Reply:
Currently the code retries on the first two rejections from `Shot.getRawBytesForClip` and rejects with its Error on the third.  I don't think we can switch to the proposed promise chain without losing the original Error from `Shot.getRawBytesForClip`.  In order for `getRawBytesPromise` to return something falsey, it needs to ignore the rejections.

Reply:
Sure, you could handle the rejection case any number of ways while still getting rid of the manual iteration counting code. We can chat about this tomorrow after standup :+1:

Reply:
Sorry, I wasn't clear: I want to reject if it's `null` but retry two times if it's a rejection.  `Shot.getRawBytesForClip` returns null when a (not {deleted,blocked,expired}) image is not found in the DB, so that's not an AWS/S3 related issue.